src/Compilers/CSharp/Test/Semantic/Semantics/LambdaTests.cs

```diff
@@ -4329,6 +4329,584 @@ static void Main()
+                //         Func<string> f3 = string? () => throw null!;
+                Diagnostic(ErrorCode.WRN_NullabilityMismatchInReturnTypeOfTargetDelegate, "string? () => throw null!").WithArguments("lambda expression", "System.Func<string>").WithLocation(10, 27),
+                // (13,31): warning CS8621: Nullability of reference types in return type of 'lambda expression' doesn't match the target delegate 'Func<S<object?>>' (possibly because of nullability attributes).
+                //         Func<S<object?>> f6 = S<object> () => throw null!;
+                Diagnostic(ErrorCode.WRN_NullabilityMismatchInReturnTypeOfTargetDelegate, "S<object> () => throw null!").WithArguments("lambda expression", "System.Func<S<object?>>").WithLocation(13, 31),
+                // (14,30): warning CS8621: Nullability of reference types in return type of 'lambda expression' doesn't match the target delegate 'Func<S<object>>' (possibly because of nullability attributes).
+                //         Func<S<object>> f7 = S<object?> () => throw null!;
+                Diagnostic(ErrorCode.WRN_NullabilityMismatchInReturnTypeOfTargetDelegate, "S<object?> () => throw null!").WithArguments("lambda expression", "System.Func<S<object>>").WithLocation(14, 30));
+        }
+
+        [Fact]
+        public void LambdaReturnType_08()
+        {
+            var source =
+@"delegate T D1<T>(ref T t);
+delegate ref T D2<T>(ref T t);
+class Program
+{
+    static void F<T>()
+    {
+        D1<T> d1;
+        D2<T> d2;
+        d1 = T (ref T t) => t;
+        d2 = T (ref T t) => t;
+        d1 = (ref T (ref T t) => ref t);
+        d2 = (ref T (ref T t) => ref t);
+    }
+}";
+            var comp = CreateCompilation(source, parseOptions: TestOptions.RegularPreview);
+            comp.VerifyDiagnostics(
+                // (10,14): error CS8918: Cannot convert lambda expression to type 'D2<T>' because the return type does not match the delegate return type
+                //         d2 = T (ref T t) => t;
+                Diagnostic(ErrorCode.ERR_CantConvAnonMethReturnType, "T (ref T t) => t").WithArguments("lambda expression", "D2<T>").WithLocation(10, 14),
+                // (11,15): error CS8918: Cannot convert lambda expression to type 'D1<T>' because the return type does not match the delegate return type
+                //         d1 = (ref T (ref T t) => ref t);
+                Diagnostic(ErrorCode.ERR_CantConvAnonMethReturnType, "ref T (ref T t) => ref t").WithArguments("lambda expression", "D1<T>").WithLocation(11, 15));
+        }
+
+        [Fact]
+        public void LambdaReturnType_09()
+        {
+            var source =
+@"using System;
+class Program
+{
+    static void Main()
+    {
+        Delegate d;
+        d = (ref void () => { });
```

Comment:
Let's add a passing test involving a void-returning lambda too.

Reply:
I couldn't find the test you added to address this comment

Reply:
I updated `DelegateTypeTests.GetLambdaData()` which is used for tests that compile and execute the lambdas.